Sources/ApodiniDeploy/ApodiniDeployInterfaceExporter.swift

```diff
@@ -0,0 +1,250 @@
+    struct ApplicationStorageKey: Apodini.StorageKey {
+        typealias Value = ApodiniDeployInterfaceExporter
+    }
+    
+    struct CollectedEndpointInfo {
+        let handlerType: HandlerTypeIdentifier
+        let endpoint: AnyEndpoint
+        let deploymentOptions: DeploymentOptions
+    }
+    
+    
+    let app: Apodini.Application
+    
+    var collectedEndpoints: [CollectedEndpointInfo] = []
+    var explicitlyCreatedDeploymentGroups: [DeploymentGroup.ID: Set<AnyHandlerIdentifier>] = [:]
+    
+    var deploymentProviderRuntime: DeploymentProviderRuntimeSupport?
+    
+    
+    public required init(_ app: Apodini.Application) {
+        self.app = app
+        app.storage.set(ApplicationStorageKey.self, to: self)
+    }
+    
+    
+    public func export<H: Handler>(_ endpoint: Endpoint<H>) {
+        if let groupId = endpoint.context.get(valueFor: DSLSpecifiedDeploymentGroupIdContextKey.self) {
+            if explicitlyCreatedDeploymentGroups[groupId] == nil {
+                explicitlyCreatedDeploymentGroups[groupId] = []
+            }
+            explicitlyCreatedDeploymentGroups[groupId]!.insert(endpoint.identifier)
+        }
+        collectedEndpoints.append(CollectedEndpointInfo(
+            handlerType: HandlerTypeIdentifier(H.self),
+            endpoint: endpoint,
+            deploymentOptions: CollectedOptions(reducing: [
+                endpoint.handler.getDeploymentOptions(),
+                endpoint.context.get(valueFor: HandlerDeploymentOptionsSyntaxNodeContextKey.self)
+            ].flatMap { $0.compactMap { $0.resolve(against: endpoint.handler) } })
+        ))
+        app.vapor.app.add(Vapor.Route(
+            method: .POST,
+            path: ["__apodini", "invoke", .constant(endpoint.identifier.rawValue)],
+            responder: InternalInvocationResponder(RHIIE: self, endpoint: endpoint),
+            requestType: InternalInvocationResponder<H>.Request.self,
+            responseType: InternalInvocationResponder<H>.Response.self
+        ))
+    }
+    
+    
+    public func finishedExporting(_ webService: WebServiceModel) {
+        do {
+            try performDeploymentStuff()
+        } catch {
+            fatalError("ugh \(error)")
+        }
+    }
+    
+    
+    private func performDeploymentStuff() throws {
+        let args = CommandLine.arguments
+        guard args.count >= 3 else {
+            return
+        }
+        
+        switch args[1] {
+        case WellKnownCLIArguments.exportWebServiceModelStructure:
+            let outputUrl = URL(fileURLWithPath: args[2])
+            do {
+                try self.exportWebServiceStructure(
+                    to: outputUrl,
+                    deploymentConfig: self.app.storage.get(ApodiniDeployConfiguration.StorageKey.self)?.config ?? .init()
+                )
+            } catch {
+                fatalError("Error exporting web service structure: \(error)")
```

Comment:
In this specific case, we end up here after launching the web service with the `WellKnownCLIArguments.exportWebServiceModelStructure` argument, which tells the web service to write its model to the file and then terminate (the web service is running with the sole purpose to perform the initial run of the semantic model builder), so I'm not sure delegating the error to the caller would be a good idea, since the web service should be terminated anyway as soon as possible after writing the model file.

I do agree with you that in general, errors should be able to be caught somewhere up the call stack, and the other branch (`WellKnownCLIArguments.launchWebServiceInstanceWithCustomConfig`) does in fact do that exact thing.

Reply:
Thanks for the explanation, sounds good to me 👍